libraries/AP_OpticalFlow/AP_OpticalFlow_MAV.cpp

```diff
@@ -0,0 +1,115 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_OpticalFlow_MAV.h"
+#include <AP_Math/edc.h>
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <utility>
+#include "OpticalFlow.h"
+#include <stdio.h>
+
+#define OPTFLOW_MAV_TIMEOUT_SEC 0.5f    // timeout after 0.5 seconds
+
+extern const AP_HAL::HAL& hal;
+
+// detect the device
+AP_OpticalFlow_MAV *AP_OpticalFlow_MAV::detect(OpticalFlow &_frontend)
+{
+    // we assume mavlink messages will be sent into this driver
+    AP_OpticalFlow_MAV *sensor = new AP_OpticalFlow_MAV(_frontend);
+    return sensor;
+}
+
+// read latest values from sensor and fill in x,y and totals.
+void AP_OpticalFlow_MAV::update(void)
+{
+    // record gyro values as long as they are being used
+    // the sanity check of dt below ensures old gyro values are not used
+    if (gyro_sum_count < 1000) {
```

Comment:
Some sort of rolling average might be better here; first time you use the data after a hiatus you're going to get ancient data.

Reply:
actually I don't think we will end up using ancient data.  In cases where it has been a long time since the last flow sensor reading the "dt" value will be very large (>0.5 seconds) and so the gyro_sum values won't be used (in the "bodyRate") and then at the bottom of update we reset the gyro_sum to zero.